devito/arguments.py

```diff
@@ -158,198 +107,350 @@ class PtrArgument(Argument):
-
-    @property
-    def value(self):
-        """ Returns a tuple (same order as rtargs) with the current value of each rtarg
-            If any of the rtargs has value None, the return value here is None.
-        """
-        child_values = tuple([i.value for i in self.rtargs])
-        return child_values if all(i is not None for i in child_values) else None
-
-    @property
-    def dtype(self):
-        """The data type of the iteration variable"""
-        return np.int32
-
-    @cached_property
-    def rtargs(self):
-        size = ScalarArgument(self.size_name, self, max)
-        start = ScalarArgument(self.start_name, self, max, 0)
-        end = ScalarArgument(self.end_name, self, max)
-        return namedtuple("RuntimeArguments", ["size", "start", "end"])(size, start, end)
-
-    # TODO: Can we do without a verify on a dimension?
-    def verify(self, value, enforce=False):
+    def __init__(self, stencils, parameters, dle_arguments):
+        self.stencils = stencils
+        self.parameters = parameters
+        self.dle_arguments = dle_arguments
+        self.argument_mapper = self._build_argument_mapper(parameters)
+        self.arguments = filter_ordered([x for x in self.argument_mapper if isinstance(x, Argument)], key=lambda x: x.name)
+        self.dims = [x for x in self.argument_mapper if isinstance(x, DimensionParameter)]
+        self.offsets = {d.end_name: v for d, v in retrieve_offsets(stencils).items()}
+
+    def handle(self, **kwargs):
+
+        user_autotune = kwargs.pop('autotune', False)
+
+        kwargs = self._offset_adjust(kwargs)
+        
+        kwargs = self._extract_children_of_composites(kwargs)
+
+        values = self._derive_values(kwargs)
+
+        # The following is only being done to update the autotune flag. The actual value derivation for the
+        # dle arguments has moved inside the above _derive_values method.
+        # TODO: Refactor so this is not required
+        dim_sizes = dict([(d.name, runtime_dim_extent(d, values)) for d in self.dimensions])
+        dle_arguments, dle_autotune = self._dle_arguments(dim_sizes)
+        
+
+        assert(self._verify(values))
+
+        arguments = OrderedDict([(k.name, v) for k, v in values.items()])
+        return arguments, user_autotune and dle_autotune
+
+    def _offset_adjust(self, kwargs):
+        for k, v in kwargs.items():
+            if k in self.offsets:
+                kwargs[k] = v + self.offsets[k]
+        return kwargs
+
+    def _build_argument_mapper(self, parameters):
+        # Pass through SymbolicFunction
+        symbolic_functions = [x for x in parameters if isinstance(x, SymbolicFunction)]
+        dimension_dependency_mapper = dict()
+        tensor_arguments = []
+        for f in symbolic_functions:
+            argument = ArgumentVisitor().visit(f)
+            tensor_arguments.append(argument)
+            for i, d in enumerate(f.indices):
+                if d not in dimension_dependency_mapper:
```

Comment:
the pythonic way of doing this sort of things is:
``dimension_dependency_mapper.setdefault(d, []).append(...)``

can we call this just ``mapper``?

Reply:
There are multiple mappers in the namespace at this point so the name is being specific about what it maps, dimensions to their dependencies.